### ISSUE TITLE
Adagrad optimizer

### ISSUE THREAD

**Issue by berryweinst** [facebookresearch/dlrm#15]
Hi,

I see in the paper that you get slightly better results with Adagrad rather than SGD, but your code supports only SGD. Why is that? 

Thanks,
Berry

**Comment by mnaumovfb** [comment]
The code can be used with any optimization algorithm. However, it is important to note that not all optimization algorithms in PyTorch support sparse gradient updates to the embedding tables. 

SGD supports sparse updates, which are enabled by setting "sparse=True" argument in the call
nn.EmbeddingBag(n, m, mode="sum", sparse=True)

Adagrad did not support sparse updates, and therefore required the entire embedding to be updated, even though most of the update vectors/elements would have been zero. This would in turn make the code less performant and require more memory than expected. That's why we have decided to only use SGD in the code.

In the future, Adagrad should support sparse updates as well, but for now if you are just interested in experimenting with different optimization algorithms and analyzing accuracy (not performance), you should be able to use Adagrad by setting the flag "sparse=False" above.

**Comment by berryweinst** [comment]
Thanks for you comment. However, I do see that Adagrad supports sparse updates when the sparse flag is set. Please see Adagrad update step code:

    def step(self, closure=None):
        """Performs a single optimization step.

        Arguments:
            closure (callable, optional): A closure that reevaluates the model
                and returns the loss.
        """
        loss = None
        if closure is not None:
            loss = closure()

        for group in self.param_groups:
            for p in group['params']:
                if p.grad is None:
                    continue

                grad = p.grad.data
                state = self.state[p]

                state['step'] += 1

                if group['weight_decay'] != 0:
                    if p.grad.data.is_sparse:
                        raise RuntimeError("weight_decay option is not compatible with sparse gradients")
                    grad = grad.add(group['weight_decay'], p.data)

                clr = group['lr'] / (1 + (state['step'] - 1) * group['lr_decay'])

                if grad.is_sparse:
                    grad = grad.coalesce()  # the update is non-linear so indices must be unique
                    grad_indices = grad._indices()
                    grad_values = grad._values()
                    size = grad.size()

                    def make_sparse(values):
                        constructor = grad.new
                        if grad_indices.dim() == 0 or values.dim() == 0:
                            return constructor().resize_as_(grad)
                        return constructor(grad_indices, values, size)
                    state['sum'].add_(make_sparse(grad_values.pow(2)))
                    std = state['sum'].sparse_mask(grad)
                    std_values = std._values().sqrt_().add_(1e-10)
                    p.data.add_(-clr, make_sparse(grad_values / std_values))
                else:
                    state['sum'].addcmul_(1, grad, grad)
                    std = state['sum'].sqrt().add_(1e-10)
                    p.data.addcdiv_(-clr, grad, std)

        return loss

**Comment by mnaumovfb** [comment]
The sparse support might have been added recently. In this case, you should be able to just use Adagrad optimization algorithm without additional modifications to EmbeddingBag.
